packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -854,7 +856,213 @@ let rewrite_jsx =
+    | _ ->
+        [%expr [%ocaml.error "server-reason-react: server functions must have a return type annotation (Js.Promise.t)"]]
+
+  let map_arguments_to_expressions ~loc args =
+    List.map
+      ~f:(fun arg ->
+        match arg with
+        | Ok (arg_label, Some arg_name, _) -> (arg_label, [%expr [%e evar ~loc arg_name]])
+        | Ok (arg_label, _, [%type: unit]) -> (arg_label, [%expr ()])
+        | Ok _ ->
+            ( Nolabel,
+              [%expr
+                [%ocaml.error
+                  "server-reason-react: invalid argument, it must have a argument with name and type annotation"]] )
+        | Error (loc, msg) -> (Nolabel, [%expr [%ocaml.error [%e estring ~loc msg]]]))
+      args
+
+  let encode_function_response ~loc ~response_expr ~core_type =
+    [%expr
+      try [%e response_expr] |> Lwt.map (fun response -> [%e response_to_json ~loc core_type [%expr response]])
+      with e -> Lwt.fail e]
+
+  let decode_arguments_vb ~loc args_to_decode =
+    List.mapi
+      ~f:(fun i (_, label, core_type) ->
+        let string_of_core_type x =
+          let f = Format.str_formatter in
+          Astlib.Pprintast.core_type f x;
+          Format.flush_str_formatter ()
+        in
+        let core_type_string = string_of_core_type core_type in
+        let of_json = make_of_json ~loc core_type [%expr args.([%e eint ~loc i])] in
+        value_binding ~loc
+          ~pat:[%pat? [%p ppat_var ~loc { txt = label; loc }]]
+          ~expr:
+            [%expr
+              try [%e of_json]
+              with _ ->
+                failwith
+                  (Printf.sprintf "server-reason-react: error on decoding argument '%s'. EXPECTED: %s, RECEIVED: %s"
+                     [%e estring ~loc label] [%e estring ~loc core_type_string]
+                     (args.([%e eint ~loc i]) |> Yojson.Basic.to_string))])
+      args_to_decode
+
+  let create_router_handler ~loc ~id ~function_name ~args ~core_type =
+    let apply_args = map_arguments_to_expressions ~loc args in
+    let response_expr = pexp_apply ~loc [%expr [%e evar ~loc function_name].call] apply_args in
+
+    let encoded_response_expr = encode_function_response ~loc ~response_expr ~core_type in
+    let args_to_decode =
+      List.filter_map
+        ~f:(fun arg ->
+          match arg with
+          | Ok (arg_label, Some arg_name, core_type) -> Some (arg_label, arg_name, core_type)
+          | Ok _ -> None
+          | Error _ -> None)
+        args
+    in
+
+    let body_expr =
+      match args_to_decode with
+      | [] -> encoded_response_expr
+      | args_to_decode ->
+          let decoded_expr = decode_arguments_vb ~loc args_to_decode in
+          pexp_let ~loc Nonrecursive decoded_expr encoded_response_expr
+    in
+    [%stri FunctionReferences.register [%e estring ~loc id] (Body (fun args -> [%e body_expr]))]
+
+  let create_server_function_contract ~loc id expression =
```

Comment:
nit: `contract` is a weird name for OCaml, since there's no interfaces, neither traits.

I know what you mean, but would suggest create_server_function_record or type